core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -1790,16 +1789,71 @@ class WorkspaceService(
       Set(WorkbenchEmail(ctx.userInfo.userEmail.value))
     }
+
+    val groupedAclUpdates: Map[WorkspaceAccessLevel, List[WorkspaceACLUpdate]] = workspaceRequest.addUsers
+      .map { aclUpdates =>
+        aclUpdates.groupBy(_.accessLevel)
+      }
+      .getOrElse(Map.empty)
+
+    def getEmailsByAccessLevel(access: WorkspaceAccessLevel): Set[WorkbenchEmail] =
+      groupedAclUpdates
+        .getOrElse(access, Set.empty)
+        .map(update => WorkbenchEmail(update.email))
+        .toSet
+
+    def getShareEmailsByAccessLevel(access: WorkspaceAccessLevel): Set[WorkbenchEmail] =
+      groupedAclUpdates
+        .getOrElse(access, Set.empty)
+        .filter(_.canShare.get)
+        .map(update => WorkbenchEmail(update.email))
+        .toSet
+
     val ownerPolicy =
-      SamWorkspacePolicyNames.owner -> SamPolicy(ownerPolicyMembership, Set.empty, Set(SamWorkspaceRoles.owner))
-    val writerPolicy = SamWorkspacePolicyNames.writer -> SamPolicy(Set.empty, Set.empty, Set(SamWorkspaceRoles.writer))
-    val readerPolicy = SamWorkspacePolicyNames.reader -> SamPolicy(Set.empty, Set.empty, Set(SamWorkspaceRoles.reader))
+      SamWorkspacePolicyNames.owner -> SamPolicy(
+        ownerPolicyMembership ++ getEmailsByAccessLevel(WorkspaceAccessLevels.Owner),
+        Set.empty,
+        Set(SamWorkspaceRoles.owner)
+      )
+    val writerPolicy =
+      SamWorkspacePolicyNames.writer -> SamPolicy(
+        getEmailsByAccessLevel(WorkspaceAccessLevels.Write),
+        Set.empty,
+        Set(SamWorkspaceRoles.writer)
+      )
+    val readerPolicy =
+      SamWorkspacePolicyNames.reader -> SamPolicy(
+        getEmailsByAccessLevel(WorkspaceAccessLevels.Read),
+        Set.empty,
+        Set(SamWorkspaceRoles.reader)
+      )
+
     val shareReaderPolicy =
-      SamWorkspacePolicyNames.shareReader -> SamPolicy(Set.empty, Set.empty, Set(SamWorkspaceRoles.shareReader))
+      SamWorkspacePolicyNames.shareReader -> SamPolicy(
+        getShareEmailsByAccessLevel(WorkspaceAccessLevels.Read),
+        Set.empty,
+        Set(SamWorkspaceRoles.shareReader)
+      )
     val shareWriterPolicy =
-      SamWorkspacePolicyNames.shareWriter -> SamPolicy(Set.empty, Set.empty, Set(SamWorkspaceRoles.shareWriter))
+      SamWorkspacePolicyNames.shareWriter -> SamPolicy(
+        getShareEmailsByAccessLevel(WorkspaceAccessLevels.Write),
+        Set.empty,
+        Set(SamWorkspaceRoles.shareWriter)
+      )
+
     val canComputePolicy =
-      SamWorkspacePolicyNames.canCompute -> SamPolicy(Set.empty, Set.empty, Set(SamWorkspaceRoles.canCompute))
+      SamWorkspacePolicyNames.canCompute -> SamPolicy(
+        workspaceRequest.addUsers
+          .map { aclUpdates =>
+            aclUpdates
+              .filter(_.canCompute.get)
+              .map(update => WorkbenchEmail(update.email))
+              .toSet
+          }
+          .getOrElse(Set.empty),
+        Set.empty,
+        Set(SamWorkspaceRoles.canCompute)
+      )
     val canCatalogPolicy =
```

Comment:
Looking at how to add users to a workspace now, there is no option for cataloging.  I'm guessing this isn't relevant for workspaces and should always be blank, or do I need to account for the possibility?

Reply:
You are correct to leave it as is. Cataloging was a feature for the old firecloud data library. I don't think it is in use anymore.